### PR TITLE
Add support for individual palettes per frame.

### PR DESCRIPTION
**Pull request by ShreksHellraiser**
Wording could be adjusted to make it clearer.

### CODE REVIEW

**Comment by ShreksHellraiser** [comment]
Should the palette key be moved from the root of the file? Instead of a palette table of palettes we instead give each frame an optional palette key that contains the palette for that frame?

**Comment by SkyTheCodeMaster** [comment]
Current discussion in `Image formats` thread in the [MCCM Guild](https://canary.discord.com/channels/477910221872824320/980612236655603752) suggest that instead of 
```lua
{
  palette = {
    {
      colour = code
    },
}
```
The following is suggested: 
```lua
{
  {
    {"image data","aaaaa0aaaa","ffffffffff"},
    palette = {
      colour = code,
    },
  }
}
``` 
The difference here is that the palette is stored inside of the frame table instead of the root palette table.
The use of the root palette table will be loaded before frames are being drawn, whereas the palette tables inside the frames are set before that specific frame is drawn.

**Comment by Ocawesome101** [comment]
i second Sky's thoughts.

**Comment by ShreksHellraiser** [comment]
I've updated `rough-ideas.md` to match this specification, and adjusted the examples to use the default first frame. The examples may want to be modified to make use of the per-frame palettes.

File: rough-ideas.md [hunk 1]
```diff
@@ -5,14 +5,15 @@ These are a rough set of fields in the `.bimg` format.
 A line is a tuple of (text, text colors, background colors) in blit format.  
 Frames are lists of lines.  
 Image files may be a single frame, or list of frames.  
+Each frame can have a `palette` key to specify a specific palette for that frame, following the rules of the base `palette` key.
```
Comment (SkyTheCodeMaster) [inline rough-ideas.md:8]: This is for *image data*, not metadata. This should be sitting down in the metadata section prepended with `(per frame)`

File: rough-ideas.md [hunk 2]
```diff
@@ -5,14 +5,15 @@ These are a rough set of fields in the `.bimg` format.
...
   
 # Metadata
 Metadata is stored at the root of the table.    
 `version`: string Version string of the format.  
 `animation`: boolean Whether the image is an animation (if so, each frame is numerically indexed).  
 (per frame) `duration`: number Duration of the frame in seconds, overrides `secondsPerFrame` for that frame.  
 `secondsPerFrame`: number Default length of each frame in seconds.  
-`palette`: table Table of color tables indexed by frame. Each color table is in format of colors api (key: color number - value: EITHER table of 3 ranged 0-1 floats for each RGB component, OR a single item which is the hex-formatted number #RRGGBB)  
+`palette`: table Palette table in format of colors api (key: color number - value: EITHER table of 3 ranged 0-1 floats for each RGB component, OR a single item which is the hex-formatted number #RRGGBB). This is optional and will serve as the starting palette for any animations or images.
```
Comment (SkyTheCodeMaster) [inline rough-ideas.md]: The optional statement isn't required, as every single one of these metadata flags (save `animation`) are required.

**Review by SkyTheCodeMaster — CHANGES_REQUESTED** [review]
Overall looks good, however the specification should be adjusted according to the comments I left.

**Review by SkyTheCodeMaster — APPROVED** [review]
Looks good to me, we should await other's opinions before merging this into the standard!

**Comment by spr2-dev** [comment]
Looks good to me, adding a palette per frame increases flexibility and that's always nice to have.
Should the `version` metadata be increased, as the first deserializers were already made?

**Comment by ShreksHellraiser** [comment]
`version`'s format hasn't been discussed yet (as far as I'm aware), so I'm not sure that there's anything *to* increment

**Comment by spr2-dev** [comment]
True, the spec isn't released yet anyway, so the edits made now will count towards the first "official" version of the format.

**Comment by SkyTheCodeMaster** [comment]
It does seem that a large number of people are in agreeance.
Merging!

**Pull request merged**